exchanges/exchange.go

```diff
@@ -1136,3 +1133,111 @@ func (e *Base) SetSaveTradeDataStatus(enabled bool) {
 		log.Debugf(log.Trade, "Set %v 'SaveTradeData' to %v", e.Name, enabled)
 	}
 }
+
+// NewEndpoints declares default and running URLs maps
+func (e *Base) NewEndpoints() *Endpoints {
+	return &Endpoints{
+		defaults: make(map[string]string),
+	}
+}
+
+// SetDefaultEndpoints declares and sets the default URLs map
+func (e *Endpoints) SetDefaultEndpoints(m map[URL]string) error {
+	for k, v := range m {
+		err := e.SetRunning(k.String(), v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SetRunning populates running URLs map
+func (e *Endpoints) SetRunning(key, val string) error {
+	e.Lock()
+	defer e.Unlock()
+	err := validateKey(key)
+	if err != nil {
+		return err
+	}
+	_, err = url.ParseRequestURI(val)
+	if err != nil {
+		log.Warn(log.ExchangeSys, fmt.Sprintf("Could not set custom URL for %s to %s. invalid URI for request.", key, val))
+		return nil
+	}
+	e.defaults[key] = val
+	return nil
+}
+
+func validateKey(keyVal string) error {
+	for x := range keyURLs {
+		if keyURLs[x].String() == keyVal {
+			return nil
+		}
+	}
+	return errors.New("keyVal invalid")
+}
+
+// GetURL gets default url from URLs map
+func (e *Endpoints) GetURL(key URL) (string, error) {
+	e.RLock()
+	defer e.RUnlock()
+	val, ok := e.defaults[key.String()]
+	if !ok {
+		return "", fmt.Errorf("no endpoint path found for the given key: %v", key)
+	}
+	return val, nil
+}
+
+// GetURLMap gets all urls for either running or default map based on the bool value supplied
+func (e *Endpoints) GetURLMap() map[string]string {
+	e.RLock()
+	var urlMap = make(map[string]string)
+	for k, v := range e.defaults {
+		urlMap[k] = v
+	}
+	e.RUnlock()
+	return urlMap
+}
+
+// FormatSymbol formats the given pair to a string suitable for exchange API requests
+func (e *Base) FormatSymbol(pair currency.Pair, assetType asset.Item) (string, error) {
+	pairFmt, err := e.GetPairFormat(assetType, true)
+	if err != nil {
+		return pair.String(), err
+	}
+	return pairFmt.Format(pair), nil
+}
+
+func (u URL) String() string {
+	switch u {
+	case RestSpot:
+		return "RestSpotURL"
+	case RestSpotSupplementary:
+		return "RestSpotSupplementaryURL"
+	case RestUSDTMargined:
+		return "RestUSDTMarginedFuturesURL"
+	case RestCoinMargined:
+		return "RestCoinMarginedFuturesURL"
+	case RestFutures:
+		return "RestFuturesURL"
+	case RestSandbox:
+		return "RestSandboxURL"
+	case RestSwap:
+		return "RestSwapURL"
+	case WebsocketSpot:
+		return "WebsocketSpotURL"
+	case WebsocketSpotSupplementary:
+		return "WebsocketSpotSupplementaryURL"
+	case ChainAnalysis:
+		return "ChainAnalysisURL"
+	case EdgeCase1:
+		return "EdgeCase1URL"
+	case EdgeCase2:
+		return "EdgeCase2URL"
+	case EdgeCase3:
+		return "EdgeCase3URL"
+	default:
+		return ""
+	}
+}
```

Comment:
Last request! The function `StoreAssetPairFormat` is adding new currency pair formats to the config such as coinmarginfutures. However, when they are initially added, the field `assetEnabled` is set to `nil`. I would like you to update the function `StoreAssetPairFormat` so that when `f.AssetEnabled == nil` that it gets set to false instead